src/Drivers/nlpPriDec_ex9_user_recourse_sparse_raja.hpp

```diff
@@ -0,0 +1,577 @@
+  {
+    auto* x_vec = const_cast<double*>(x);
+
+    auto& resmgr = umpire::ResourceManager::getInstance();
+    umpire::Allocator allocator = resmgr.getAllocator(mem_space_);
+    if(x_==NULL) {
+      x_ = static_cast<double*>(allocator.allocate(nx_ * sizeof(double)));
+    }
+    auto* str = allocator.getAllocationStrategy();
+    resmgr.registerAllocation(x_vec, {x_vec,nx_*sizeof(double),str});
+    resmgr.copy(x_, x_vec);  
+    resmgr.deregisterAllocation(x_vec); 
+  }
+
+  /// Set the "sample" vector \xi
+  void set_center(const double *xi)
+  {
+    auto* xi_vec = const_cast<double*>(xi);
+    auto& resmgr = umpire::ResourceManager::getInstance();
+    umpire::Allocator allocator = resmgr.getAllocator(mem_space_);
+    
+    if(xi_ == NULL) {
+      xi_ = static_cast<double*>(allocator.allocate(nS_ * sizeof(double)));
+    }
+    auto* str = allocator.getAllocationStrategy();
+    resmgr.registerAllocation(xi_vec, {xi_vec,nS_*sizeof(double),str});
+    resmgr.copy(xi_, xi_vec);
+    resmgr.deregisterAllocation(xi_vec); 
+  }
+
+  bool get_prob_sizes(size_type& n, size_type& m)
+  {
+    n = ny_;
+    m = ny_; 
+    return true; 
+  }
+
+  bool get_vars_info(const size_type& n, double *xlow, double* xupp, NonlinearityType* type)
+  {
+    // y_1 bounded
+    xlow[0] = 0.;
+   
+    xupp[0] = 1e20;
+    RAJA::forall<ex9_raja_exec>(RAJA::RangeSegment(1, n),
+    RAJA_LAMBDA(RAJA::Index_type i)
+    {
+      xlow[i] = -1e+20;
+    });
+
+    RAJA::forall<ex9_raja_exec>(RAJA::RangeSegment(1, n),
+    RAJA_LAMBDA(RAJA::Index_type i)
+    {
+      xupp[i] = 1e+20;
+    });
+
+    RAJA::forall<RAJA::loop_exec>(RAJA::RangeSegment(0, n),
+    [=] (RAJA::Index_type i)
+    {
+      type[i] = hiopNonlinear;
+    });
+    //assert(false && "not implemented");
+    return true;
+  }
+
+  bool get_cons_info(const size_type& m, double* clow, double* cupp, NonlinearityType* type)
+  {
+    assert(m == ny_);
+    RAJA::forall<ex9_raja_exec>(RAJA::RangeSegment(0, ny_-1),
+    RAJA_LAMBDA(RAJA::Index_type i)
+    {
+      clow[i] = 0.;
+      cupp[i] = 1e20;
+    });
+
+    //clow[ny_-1] = 0.;
+    clow[ny_-1] = 1.; 
+    cupp[ny_-1] = 1e20;
+    return true;
+  }
+
+  bool get_sparse_blocks_info(size_type& nx, 
+                              size_type& nnz_sparse_Jaceq,
+                              size_type& nnz_sparse_Jacineq,
+                              size_type& nnz_sparse_Hess_Lagr) 
+  {
+    nx = ny_;
+    assert(nx>0);
+
+    nnz_sparse_Jaceq = 0;
+    nnz_sparse_Jacineq = ny_+(ny_-1)*2;
+    
+    nnz_sparse_Hess_Lagr = ny_;  //Lagrangian
+    return true;
+  }
+
+  bool eval_f(const size_type& n, const double* x, bool new_x, double& obj_value)
+  {
+    assert(ny_==n);
+    obj_value = 0.;
+    RAJA::ReduceSum<ex9_raja_reduce, double> aux(0); //why do we need reducesum?
+    RAJA::forall<ex9_raja_exec>(RAJA::RangeSegment(0, n),
+      RAJA_LAMBDA(RAJA::Index_type i)
+      {
+        aux += (x[i]-x_[i])*(x[i]-x_[i]);
+      });
```

Comment:
we need to use reducesum here. Otherwise we will have race condition and aux may return a wrong value.

Reply:
Thanks. Makes sense. I checked other cases and they are all `some_vector[i] += something`, not a scalar. So should be fine.